[Sketcher] Reducing cognitive complexity

AjinkyaDahale opened the pull request:
Done as [a grant by the FPA](https://github.com/FreeCAD/FPA-grant-proposals/issues/6).

3x380V commented:
Now looking at whole `SketchObject::getPoint` and before you spend too much time here... The template you are creating can be used at most two times or you need also add emulateCCW parameter and whoever is going to read that after your change needs to look at template definition. What about simple:
```
    } else if (geo->is<Part::GeomArcOfCircle>()) {
        const Part::GeomArcOfCircle *aoc = static_cast<const Part::GeomArcOfCircle*>(geo);
        switch (PosId) {
            case PointPos::start:
                return aoc->getStartPoint(/*emulateCCW=*/true);
            case PointPos::end:
                return aoc->getEndPoint(/*emulateCCW=*/true);
            case PointPos::mid:
                return aoc->getCenter();
        }
    } else if (geo->is<Part::GeomArcOfEllipse>()) {
```
This is exactly what switch is for, but this still is just shuffling around. You might come with looping around class list using lambdas, however doubts about how much is that improvement remain.

AjinkyaDahale commented:
> Now looking at whole `SketchObject::getPoint` and before you spend too much time here... The template you are creating can be used at most two times or you need also add emulateCCW parameter and whoever is going to read that after your change needs to look at template definition.

I'm looking to make it such that this template can be used in other places including split, trim, extend, move etc. However, this will need some means of passing the additional parameters, which I'm trying to figure out.

> What about simple:
> 
> This is exactly what switch is for, but this still is just shuffling around. You might come with looping around class list using lambdas, however doubts about how much is that improvement remain.

Simply replacing a small example with switch-case will not reduce the cognitive complexity, at least not as much as extracting the per geometry logic would.

FlachyJoe commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -1019,68 +1020,128 @@ int SketchObject::movePoint(int GeoId, PointPos PosId, const Base::Vector3d& toP
     return lastSolverStatus;
 }
 
+template <>
```
These function doesn't depends of SketchObject, shouldn't they move to a SketchUtil.cpp (or whatever we name it)?

AjinkyaDahale replied:
Maybe. One step at a time. However, note that this is a very specific function taking a sketcher-specific function and getting the point based on geometry of the object (which is obtained through Part and ultimately OCCT). Such functionality has been part of SketchObject (with `Sketch` handling the functionality connected to planeGCS).

PaddleStroke replied:
+1 though this was what I attempted to do in https://github.com/FreeCAD/FreeCAD/pull/10721 and there was a roadblock which is that very annoyingly utils functions cannot be shared between GUI and APP.

To be shared you currently need to wrap them in a class object. And at this point then why not using the sketch object class...

Another alternative would be to have the same Util file in both App and Gui and keep them in sync.

hyarion replied:
The problem back then was most likely that the functions was never exported.

FlachyJoe commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -1019,68 +1020,128 @@ int SketchObject::movePoint(int GeoId, PointPos PosId, const Base::Vector3d& toP
-            return bsp->getEndPoint();
+        return getPointForGeometry<Part::GeomPoint>(static_cast<const Part::GeomPoint*>(geo), PosId);
+    }
+    else if (geo->is<Part::GeomLineSegment>()) {
```
Shouldn't we remove all the tests and use the template argument deduction?

AjinkyaDahale replied:
Is there a way to get the `Part::GeomLineSegment` part out? This kind of test is happening because 1) we are given a `Part::Geometry`; 2) it is not easily obtainable what subclass of `Part::Geometry` it is.

I do picture abstracting this check itself through templates so that it can be used in other places like `trim`, `split`, `extend`, `exposeInternalGeometry` etc.

FlachyJoe replied:
You're right, test is needed. But I think template use unnecessarily heavy and overload sufficient.
```cpp
Base::Vector3d SketchObject::getPointForGeometry(const Part::Geometry *geo, PointPos PosId){
    if (geo->is<Part::GeomPoint>()) {
        return getPointForGeometry(static_cast<const Part::GeomPoint*>(geo), PosId);
    }
    else if (geo->is<Part::GeomLineSegment>()) {
        return getPointForGeometry(static_cast<const Part::GeomLineSegment*>(geo), PosId);
    }
    // ...
}
```

AjinkyaDahale replied:
Indeed. This was pointed out to me in some private channels as well, and most likely I will change this behavior sooner than later.

Overall, this is my plan:
```cpp
Base::Vector3d SketchObject::getPoint(const Part::Geometry *geo, PointPos PosId)
{
    template <class geomType>
        auto singleArgFunction = [&GeoId, &point](T* geo) {
            return getPointForGeometry<geomType>(GeoId, point);
        };

    return performActionByGeomType<Base::Vector3d, singleArgFunction>(geo);
}

int SketchObject::trim(int GeoId, const Base::Vector3d& point)
{
    template <class geomType>
        auto singleArgFunction = [&GeoId, &point](T* geo) {
            return trimByGeomType<geomType>(GeoId, point);
        };

    return performActionByGeomType<int, singleArgFunction>(getGeometry(GeoId));
}

template <returnType, singleArgFunction>
returnType SketchObject::performActionByGeomType(Part::Geometry* geo)
{
    if (geo->is<Part::GeomPoint>()) {
        return singleArgFunction<Part::GeomPoint>(static_cast<Part::GeomPoint*>(geo));
    }
    else if (geo->is<Part::GeomLineSegment>()) {
        return singleArgFunction<Part::GeomLineSegment>(static_cast<const Part::GeomLineSegment*>(geo));
    }
    // ... other supported cases

    return singleArgFunction<Part::Geometry>(geo);
}
```

AjinkyaDahale committed:
[Sketcher][DO NOT MERGE] Refactor `SketchObject:validateExternalLinks()`

AjinkyaDahale committed:
[Sketcher][DO NOT MERGE] Refactor some loops for understandability

DO NOT MERGE since tests are not added.

Changes include:
1. Modernize `for` loops with range whenever possible.
2. Flip `if` statements for a "flatter" flow.
3. Use internal functions for repeated identical tasks.

AjinkyaDahale committed:
[Sketcher][DO NOT MERGE] Refactor based on loops and extraction into functions

May contain some untested changes

AjinkyaDahale committed:
[Sketcher][DO NOT MERGE] Refactor `SketchObject` some more

AjinkyaDahale committed:
[Sketcher][DO NOT MERGE] Refactor `SketchObject::addCopy()`

AjinkyaDahale committed:
[Sketcher] Use range-`for` loop where straightforward

AjinkyaDahale committed:
[Sketcher] Move some checks to `Constraint`

AjinkyaDahale committed:
[Sketcher][DO NOT MERGE] Refactor `delExternalPrivate()`

DO NOT MERGE since tests not added.

AjinkyaDahale committed:
[Sketcher] Refactor `removeAxesAlignment()`

Use switch case for more readability, and move some if and for indentations

AjinkyaDahale committed:
[Sketcher] Refactor `addSymmetric` and `getSymmetric`

Rearrange some if-else and use `std::find_if` instead of looping.

AjinkyaDahale committed:
[Sketcher] Refactor `SketchObject::migrateSketch()`

Manipulate some if-else statements, for loops, and possibly replace them with
std algorithms.

AjinkyaDahale committed:
[Sketcher] Refactor some if-else statements in SketchObject

AjinkyaDahale committed:
[Sketcher] Refactor `SketchObject::validateExpression()`

AjinkyaDahale committed:
[Sketcher] Refactor `addCopy`

if and for rearrangement for readability.

AjinkyaDahale committed:
[Sketcher] Refactor `SketchObject::onChanged()`

Break into pieces depending on which property has been passed.

AjinkyaDahale committed:
[Sketcher] Refactor `SketchObject::updateGeometryRefs()`

AjinkyaDahale committed:
[Sketcher][WIP] Refactor `getSymmetric` and `addSymmetric`

AjinkyaDahale committed:
[Sketcher] Use `auto` where possible

AjinkyaDahale committed:
[Sketcher] Fix issue in `replaceGeometries` when more old than new

AjinkyaDahale committed:
[Sketcher][test][WIP] Add tests for `replaceGeometries()`

AjinkyaDahale committed:
[Sketcher] Use `replaceGeometries()` in `join()`

AjinkyaDahale committed:
Sketcher: [WIP] Refactor `Sketch::addConstraint()`

AjinkyaDahale committed:
Sketcher: [WIP] Refactor `Sketch::calculateDependentParametersElements()`

AjinkyaDahale committed:
Sketcher: [WIP] Refactor `Sketch::getBlockedGeometry`

AjinkyaDahale committed:
Sketcher: [WIP] Refactor tangent and angle-via-point in `Sketch.cpp`

AjinkyaDahale committed:
Sketcher: [WIP] Refactor `Sketch::addEqualConstraint()`

AjinkyaDahale committed:
Sketcher: [WIP] Refactor `Sketch::internalSolve()`

AjinkyaDahale committed:
Sketcher: [WIP] Refactor `Sketch::initMove()`

AjinkyaDahale committed:
Sketcher: [WIP] Refactor `Sketch::MoveGeometries()`

AjinkyaDahale committed:
Sketcher: [WIP] Refactor `Sketch::calculateDependentParametersElements()`

AjinkyaDahale committed:
Sketcher: [WIP] Refactor `Sketch::getBlockedGeometry`

AjinkyaDahale committed:
Sketcher: [WIP] Refactor `Sketch::addConstraint()`

AjinkyaDahale committed:
Sketcher: Refactor `SketchObject::addCopy()`

AjinkyaDahale committed:
Sketcher: Refactor `SketchObject::addCopy`

Move nested blocks into internal functions

AjinkyaDahale committed:
Sketcher: Refactor `SketchObject::addSymmetric()`

Move nested blocks into internal functions

AjinkyaDahale committed:
Sketcher: Refactor `getSymmetric()`

AjinkyaDahale committed:
Sketcher: Refactor (aggressively) `SketchObject::getSymmetric()`

Use `Part::Geometry::mirror`.

AjinkyaDahale committed:
Sketcher: Refactor (aggressively) `SketchObject::getCopy()`

Use existing method `Part::Geometry::translate()`.

AjinkyaDahale committed:
Sketcher: Refactor `TaskSketcherElements`

AjinkyaDahale committed:
Sketcher: Refactor updateGeometryRefs

AjinkyaDahale commented:
I'm pucking at it and regularly rebasing. There's a PR named "...stage-5" which will be about as much as I can prepare for 1.1.0.
